CRM-20958 - Track creation+modification times for activities+cases

totten opened the pull request:
Overview
----------------------------------------
In developing workflows, reports, and UIs for activities and cases, it is useful to sort and filter based based on when a record was created or when a record was last modified.


Before
----------------------------------------

 * The tables `civicrm_activity` and `civicrm_case` do __not__ have the timestamp columns `created_date` and `modified_date`.
 * The table `civicrm_log` includes *some* records to indicate creation/modification time:
   * It __does__ include records for activities which have been created or updated via BAO. (Notwithstanding the possibility of local-admin pruning.)
   * It __may not__ include records for sample activities created during installation or others created via SQL.
   * It __does not__ include records for cases. (According to the original design philosophy, all changes to cases would be reflected in the activities, so I suppose `civicrm_log` would have been ~~redundant~~ low-priority.)
   * The timestamps stored in `civicrm_log.modified_date` are not entirely reliable -- they are stored as `DATETIME` without any adjustments for timezone. This is a pre-existing and systemic problem with fields being flagged as `DATETIME` instead of `TIMESTAMP`.

After
----------------------------------------

 * The tables `civicrm_activity` and `civicrm_case` __do__ have the timestamp columns `created_date` and `modified_date`.
   * These fields can be __read__ via APIv3/BAO/DAO/SQL. They cannot be directly __modified__ via APIv3/BAO.
   * These fields are automatically __initialized__ and __updated__ whenever one:
     * Creates or updates an `Activity` via APIv3/BAO/DAO/SQL.
     * Creates or updates a `Case` via APIv3/BAO/DAO/SQL.
     * Creates or updates an `Activity` via "Contacts => New Activity", "View Contact => Actions", or "View Contact => Activities". (This stems from SQL triggers, but it's been tested separately.)
     * Creates or updates a `Case` via "Cases => New Case" or "Manage Cases" UI. (This stems from SQL triggers, but it's been tested separately.)
 * The system status check displays __NOTICEs__ if...
   * There are any case or activity records which do not have `created_date` or `modified_date`.
   * Any columns are marked as `DATETIME` instead of a more sensible `TIMESTAMP`. (This addresses both `civicrm_log.modified_date` and the pre-existing discrepancies left over from [CRM-9683](https://issues.civicrm.org/jira/browse/CRM-9683)).
 * Both __NOTICEs__ suggest installation of the new (experimental) extension, [doctorwhen](https://github.com/civicrm/org.civicrm.doctorwhen), to handle migration+cleanup.
 
![screen shot 2017-07-31 at 8 45 53 pm](https://user-images.githubusercontent.com/1336047/28808501-54f8c73c-7631-11e7-88a4-1897bcb6fc47.png)

If the admin installs `doctorwhen`, then a new item appears in "Administer => Doctor When":
![screen shot 2017-07-31 at 8 59 33 pm](https://user-images.githubusercontent.com/1336047/28808727-340c55be-7633-11e7-9338-156c0086102c.png)


Technical Details
----------------------------------------
 * For initializing/maintaining the fields `{civicrm_activity,civicrm_case}.{created_date,modified_date}`, there are several SQL triggers. The configuration closely parallels `civicrm_contact`:
   * The SQL column definition for `modified_date` specifies `ON UPDATE CURRENT_TIMESTAMP`). (Note: MySQL only allows one column to be flagged with a default value of `CURRENT_TIMESTAMP`.)
   * The services `civi.activity.triggers` and `civi.case.triggers` install SQL triggers which initialize the `created_date`.
   * The services `civi.activity.triggers` and `civi.case.triggers` install SQL triggers which update `modified_date` whenever custom-data changes.
   * The service `civi.case.staticTriggers` installs SQL triggers which update the `civicrm_case.modified_date` whenever a related activity is created, modified, or deleted.
 * The list of SQL columns to consider migrating is tracked in `CRM_Utils_Check_Component_Timestamps`.

Comments
----------------------------------------
 * The JIRA ticket [CRM-20958](https://issues.civicrm.org/jira/browse/CRM-20958) includes a lengthy list of acceptance criteria.
 * This PR description has been updated to reflect issues that were discussed during development.


---

 * [CRM-20958: Data model: Track creation+modification times for activities+cases](https://issues.civicrm.org/jira/browse/CRM-20958)

totten commented on CRM/Upgrade/Incremental/php/FourSeven.php:
```diff
@@ -421,6 +421,90 @@ public function upgrade_4_7_24($rev) {
+
+      $title = sprintf('CRM-20958 - Compute civicrm_activity.created_date from civicrm_log (%d => %d)', $startId, $endId);
+      $sql = 'UPDATE civicrm_activity
+       SET created_date = (SELECT MIN(l.modified_date) FROM civicrm_log l WHERE l.entity_table ="civicrm_activity" AND civicrm_activity.id = l.entity_id)
```
One thing I'm not certain about... this code initializes `civicrm_activity.created_date` (etal) using `civicrm_log.modified_date`, but they have different data-types (`TIMESTAMP` vs `DATETIME`, respectively).

@eileenmcnaughton @seamuslee001 To my recollection, `DATETIME` and `TIMESTAMP` handle timezones differently, so wondering if there's some kind of filtering I should do here to adjust for timezone.

However, my gut says that storing `civicrm_log.modified_date` as `DATETIME` is actually a subtle dirtiness/corruption. Ex: [This](https://github.com/civicrm/civicrm-core/blob/4.7.22/CRM/Activity/BAO/Activity.php#L633-L640) suggests that `civicrm_log.modified_date` would be initialized based on `date('YmdHis')` (ie the user's *perceived timestamp* based on active PHP timezone setting) rather than a *canonical timestamp*. But different `civicrm_log` records can be created by different people/sessions/timezones, so the raw data in `civicrm_log` wouldn't be reliably articulated in any one timezone.

This wouldn't really the fault of my issue/patch -- if the theory is true, you'd expect to see idiosyncrasies in the display of this data. For migration purposes, somewhere you'd have to bite the bullet and accept that timestamps would be wrong +/- a few hours?

totten replied:
Confirmed:
 * Open a new browser. Login to `d46` as `admin`. Set your timezone to `America/New_York`. Edit contact 139. Observe "Change Log" has new entry with timestamp `July 26th, 2017 7:29 PM`.
 * Open a new browser. Login to `d46` as `demo`. Set your timezone to `America/Los_Angeles`. Edit contact 139. Observe "Change Log" has new entry with timestamp `July 26th, 2017 4:30 PM`. (In reality, this is 1 minute later.)

But the changelog (as viewed by either user) shows weird data: `admin`'s change actually happened **_1 minute before_** `demo`'s change, but it looks like it happened _**179 minutes after**_.

<img width="727" alt="screen shot 2017-07-26 at 4 34 29 pm" src="https://user-images.githubusercontent.com/1336047/28648107-53d619c8-7220-11e7-85a3-03efa4d30c2b.png">

seamuslee001 commented:
Yep so timestamp stores everything in UTC and does conversion to UTC if connection is in another Timezone. Date time is just what ever time is sent to it from the client

xurizaemon commented:
CRM-9683 seems relevant to all this - same basic problem, over in CiviMail?

seamuslee001 commented:
I thought that CiviMail fields had been converted to timestamp already

seamuslee001 commented:
Maybe that was only on the Fuzion branch of things

eileenmcnaughton commented:
@seamuslee001 we have converted some fields on new installs but not on existing due to nervousness about the - the fundamental thing I believe it when you convert you should be in the most accurate tz possible

seamuslee001 commented:
@eileenmcnaughton for some reason i thought the CiviMail was already converted, maybe AUG / fuzion thing

eileenmcnaughton commented:
@seamuslee001 yes, converted on all Fuzion customers & any site that wishes to convert can do so with a simple ALTER TABLE mysql statement. Also converted for new installs - but we have not figured out what to do with existing ones (mostly because the timestamp will be relative to the tz used when converting & there is some room for debate there)

eileenmcnaughton commented:
(there is a case to be made for providing a change over routine that people can run through UI or cli depending which is more appropriate)

totten commented:
> (there is a case to be made for providing a change over routine that people can run through UI or cli depending which is more appropriate)

Yeah, this is my main question at this point. 

On one hand... this PR seems to be working for me. The schema is created; the data migrates according to the evil scenario described in the commit-notes; the API seems to return and sort the fields; the fields for cases+activities are updated when using "Manage Case" UI; there's new test coverage; and the existing tests are now passing.

On the other hand... it might be a safer phase-in process if we only add the new columns -- and put the other stuff (the triggers and migration) behind some kind of opt-in.

eileenmcnaughton commented:
I think the triggers should be uncontroversial - since they are also in the contact schema. The question is more the migration. On balance I probably would, since that was done for the civicrm_contact table. But if you were going to leave something off I would limit it to the migration

totten commented:
@eileenmcnaughton @xurizaemon @seamuslee001  I've moved the migration logic to an extension https://github.com/civicrm/org.civicrm.doctorwhen and put a status check in `CRM_Utils_Check_Component_Timestamps`. Additionally, since CRM-9683 sorta left the transition from `DATETIME` to `TIMESTAMP` in a netherrealm, and since `doctorwhen` is labelled "experimental", it also  includes checkboxes for migrating those columns.

eileenmcnaughton commented:
cool  - I wonder how people would find out about that if they wanted to switch

totten commented:
CC @demeritcowboy - you might find this PR interesting.

seamuslee001 commented:
makes sense to me

demeritcowboy commented:
Thanks for the notice. Hmm, DoctorWhen - Timezones and Relative Dates in Civi?

Don't think it makes a difference here but noting the original use of civicrm_log wasn't entirely redundant - for example to see an activity's original creation date you still needed to look in civicrm_log.

It also probably doesn't make a difference but since the opening paragraph mentions it as the rationale, noting that at least in my experience Activity Date (activity_date_time) is usually the field used for sorting/reporting, and it allows users to have some control over how things sort without messing up the "system" dates like creation date. The system dates are usually just used during investigation. Cases and some of their activities are often allocated to certain reporting periods via the activity date, similar to financial reporting. But none of that prevents making changes for create/modify.

davialexandre commented on tests/phpunit/api/v3/ActivityTest.php:
```diff
@@ -447,6 +447,11 @@ public function testActivityCreateExample() {
     require_once 'api/v3/examples/Activity/Create.php';
     $result = activity_create_example();
     $expectedResult = activity_create_expectedresult();
+    // Mask timestamp fields. Too variable.
```
Sorry, but I'm afraid I don't understand this comment. Is is because the timestamps will change every time you run the test?

totten replied:
Yup, pretty much. I'll rephrase it.

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -165,6 +165,9 @@ public function testCaseUpdate() {
     $result = $this->callAPISuccessGetSingle('Case', array('case_id' => $id));
+    $this->assertTrue($result['modified_date'] >= $case['modified_date']);
+    unset($result['modified_date']);
+    unset($case['modified_date']);
```
What about leaving a comment here to explain why the unsets are necessary?

totten replied:
Done

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['modified_date']);
+    $this->assertApproxEquals(strtotime($activity_1['created_date']), strtotime($activity_1['created_date']), 2);
+
+    usleep(1.5 * 1000000);
```
Why is this necessary?

totten replied:
It ensures that subsequent modifications will have a new timestamp (greater than the `created_date` or previous modifications).

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_1['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_1['modified_date']);
+    $this->assertApproxEquals(strtotime($case_1['created_date']), strtotime($case_1['created_date']), 2);
```
You're comparing the created_date with itself here. Is this correct?

totten replied:
Thanks, I'll fix that.

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
+    ));
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['created_date']);
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $activity_1['modified_date']);
+    $this->assertApproxEquals(strtotime($activity_1['created_date']), strtotime($activity_1['created_date']), 2);
```
Here too

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
+
+    $this->assertEquals($activity_1['created_date'], $activity_2['created_date']);
+    $this->assertNotEquals($activity_1['modified_date'], $activity_2['modified_date']);
+    $this->assertTrue(strtotime($activity_1['modified_date']) < strtotime($activity_2['modified_date']),
```
What about using `assertLessThan()` instead of `assertTrue()`?

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -732,4 +735,52 @@ public function testCaseMerge() {
+    $this->assertRegExp(';^\d\d\d\d-\d\d-\d\d \d\d:\d\d;', $case_2['modified_date']);
+    $this->assertEquals($case_1['created_date'], $case_2['created_date']);
+    $this->assertNotEquals($case_2['created_date'], $case_2['modified_date']);
+  }
```
I must confess it was a bit hard for me to read this test and understand what was going on. I feel it would have been easier if it was split into smaller tests, each one having a descriptive name of what is being tested. Examples:
- `testCreatedDateShouldBeTheSameAsModifiedDateWhenACaseIsCreated`
- `testCreatedDatedIsNotUpdatedDuringWhenACaseIsUpdated`
- `testModifiedDateShouldBeUpdatedWhenACaseIsUpdated`
- `testCreatedDateAndModifiedDateCannotBeManuallyChanged`
- `testDateFormatOfCreatedDateAndModifiedDateShouldBeCorrect`

Note that the examples have some scenarios that were not even covered by tests. In addition to that, do you think it would make sense to also have tests for the BAO (that is, check that the same rules will work when you do something like `Case::create()`?

davialexandre commented on CRM/Upgrade/Incremental/php/FourSeven.php:
```diff
@@ -1197,4 +1217,24 @@ protected function checkImageUploadDir() {
+   *
+   * @see CRM_Core_DAO::executeQuery
+   */
+  public static function task_executeQuery($ctx, $sql, $vars) {
```
Maybe I missed it, but it doesn't look like you're using this anywhere. Are you?

totten replied:
Valid point. It was a helper to support the migration task, but then the "migration task" migrated itself off to `doctorwhen`, so we don't really need this straggler.

OTOH, it is a handy straggler...

totten commented on CRM/Utils/Check/Component/Timestamps.php:
```diff
@@ -0,0 +1,170 @@
+  /**
+   * Check that the timestamp columns are populated. (CRM-20958)
+   *
+   * @return array<CRM_Utils_Check_Message>
```
It means "an array in which the elements are instances of class `CRM_Utils_Check_Message`".

Personally, I go a bit bonkers reading `@return array` because it doesn't really sanything.

AFAIK, `php.net` hasn't blessed a notation for array typing, but there a few references for comparison:
 * It's similar to Java's notation for generics (eg `Vector<Message>`, `HashMap<String,Message>`).
 * It matches [Hack's array notation](https://docs.hhvm.com/hack/collections/introduction). (Hack is PHP derivative with stronger typing.)
 * PHP-FIG has  [a draft spec](https://github.com/php-fig/fig-standards/blob/master/proposed/phpdoc.md#abnf) for docblocks. It mentions both C-style arrays (`Message[]`) as well Java-style generics (`Collection<Message>`).
 * In my local copy of PHPStorm, it provides drilldown for C-style arrays (`Message[]`) but not Java-style generics.
 * I'm not particularly consistent. Sometimes it seems meaningful to convey even more information, so I use really verbose descriptor like:

```
  @return array
     Array(string $msgId => CRM_Utils_Check_Message $msg).
```

I guess standardizing/documenting/cleaning-up would be good, but that should be probably be a separate project.

davialexandre replied:
I believe that most projects use the notation defined by http://phpdoc.org, which uses the `CRM_Utils_Check_Message[]`. Given that the PHP-FIG draft is a derivative of that phpdoc and that most projects already use the C-style arrays, I think safe enough to assume this is the "right" way to do it.

Sometimes I prefer a more verbose description too

totten commented on CRM/Utils/Check/Component/Timestamps.php:
```diff
@@ -0,0 +1,170 @@
+   */
+  public function checkSchema() {
+    $problems = array();
+    foreach (self::getConvertedTimestamps() as $tgt) {
```
I'll rename it to `$target`.

totten commented on Civi/Core/SqlTriggers/StaticTriggers.php:
```diff
@@ -0,0 +1,124 @@
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2017
+ */
+class StaticTriggers {
```
Re: `SqlTriggers` namespace -- Agree that should be singular. Pretty much all namespaces can be equally described as singular or plural, and it's nice to have consistent style.

Re: `StaticTriggers` class -- I don't think the singular/plural issue as a matter of style here -- it's really a semantic difference. `SqlTrigger` would just be one object managing one trigger. But if one object manages multiple triggers, then you'd say `SqlTriggers` or `SqlTriggerSet` or `SqlTriggerList` or `SqlTriggerArray` or `SqlTriggerGroup`. Personally, I prefer for the simplest construction that's accurate.

In this case, this instance of `StaticTriggers` contains three triggers that serve a common goal. (The goal is to update `civicrm_case.modified_date` whenever a related activity is inserted, updated, or deleted.)

davialexandre replied:
I can't say that I agree, but I don't have a better name to suggest. Since you think it's good enough, I won't argue :)

totten commented on Civi/Core/SqlTriggers/StaticTriggers.php:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  /**
+   * Get a list of triggers for the contact table.
```
Yup, will fix.

totten commented on Civi/Core/SqlTriggers/StaticTriggers.php:
```diff
@@ -0,0 +1,124 @@
+   * @see \CRM_Core_DAO::triggerRebuild
+   * @see http://issues.civicrm.org/jira/browse/CRM-10554
+   *
+   * @param $info
```
It's a good question, but the answer is to drill-down on `@see \CRM_Utils_Hook::triggerInfo`. I don't want to reproduce the hook spec here. Will put another comment directing folks there.

totten commented on Civi/Core/SqlTriggers/StaticTriggers.php:
```diff
@@ -0,0 +1,124 @@
+   * @see http://issues.civicrm.org/jira/browse/CRM-10554
+   *
+   * @param $info
+   * @param null $tableFilter
```
It's a good question, but the answer is to drill-down on `@see \CRM_Utils_Hook::triggerInfo`. I don't want to reproduce the hook spec here. Will put another comment directing folks there.

totten commented on Civi/Core/SqlTriggers/TimestampTriggers.php:
```diff
@@ -0,0 +1,334 @@
+      //      $currentVer = CRM_Core_BAO_Domain::version(TRUE);
+      //      if (version_compare($currentVer, '4.3.alpha1') < 0) {
+      //        return;
+      //      }
```
Analytically, you're right, the commented code is superfluous and may confuse future readers. Emotionally, based on some of the history, I wanted to keep the example handy for a month. But it's not that important. Reworking the comments.

davialexandre commented on tests/phpunit/api/v3/CaseTest.php:
```diff
@@ -163,8 +163,13 @@ public function testCaseUpdate() {
+    // Verify that updated case is equal to the original with new subject.
     $result = $this->callAPISuccessGetSingle('Case', array('case_id' => $id));
+    // Modification dates are likely to differ by 0-2 sec. Check manually.
+    $this->assertTrue($result['modified_date'] >= $case['modified_date']);
```
I saw you accepted my suggestion to use `assertLessThan()` so maybe can you use `assertGreaterThanOrEqual` here, especially because there's no custom message for this assertion

totten commented:
@agh1 @MegaphoneJon @Stoob  - I think you guys are pretty thoughtful about messaging. Could you comment on how well these screens present the matter of transitioning/cleaning-up dates/times? See the screenshots and [hyperlink](https://github.com/civicrm/org.civicrm.doctorwhen) from the PR description.

In years past, we might have tried to address the change/cleanup of time data with an automated upgrade script, and then there'd be fall-out when the script doesn't anticipate some issue... which is fairly likely because no one has access to the kind of context/tools/datasets needed to ensure a change works everywhere. In theory, the RC period helps, but I'm feeling a little untrusting there. So this PR tries to break it into smaller steps -- eg make the smallest change possible, show a notification, say "Don't panic", and then link to experimental docs/tools for the rest.

The in-app messages will probably set the tone+reactions. Most of the folks on this ticket have already formed some perspective (due historical issues) -- so I wanted to ping you guys for a fresh perspective on the messaging/transition-path.

totten commented:
Following up on my questions about messaging... I escalated to `civicrm-dev` and moved half of the messaging to separate/smaller JIRA issue and PR (CRM-21079, #10874), which reduced the scope of the current PR.

Also, just rebased to handle a merge-conflict.
